File: src/problem_292.py
# Problem 292: https://projecteuler.net/problem=292

from math import gcd, isqrt


PERIMETER = 120
SPAN = 2 * PERIMETER + 1
LAYER = SPAN * SPAN


def state(x, y, p):
    return p * LAYER + (x + PERIMETER) * SPAN + (y + PERIMETER)


def edge_groups(limit):
    by_direction = {}
    opposite_edges = 0

    for x in range(-limit, limit + 1):
        x2 = x * x
        for y in range(-limit, limit + 1):
            if x == 0 and y == 0:
                continue

            sq = x2 + y * y
            length = isqrt(sq)
            if length * length != sq or length > limit:
                continue

            if 2 * length <= limit:
                opposite_edges += 1

            g = gcd(abs(x), abs(y))
            direction = (x // g, y // g)
            delta = length * LAYER + x * SPAN + y
            by_direction.setdefault(direction, []).append((length, delta))

    return [sorted(by_direction[k]) for k in sorted(by_direction)], opposite_edges // 2


def count(limit):
    groups, opposite_pairs = edge_groups(limit)
    ways = {state(0, 0, 0): 1}

    for group in groups:
        nxt = ways.copy()
        get = nxt.get
        for k, nways in ways.items():
            left = limit - k // LAYER
            for length, delta in group:
                if length > left:
                    break
                nk = k + delta
                nxt[nk] = get(nk, 0) + nways
        ways = nxt

    closed = sum(ways.get(state(0, 0, p), 0) for p in range(limit + 1))
    return closed - 1 - opposite_pairs


def solve():
    return count(PERIMETER)


if __name__ == "__main__":
    print(solve())

File: src/problem_292.go
// Problem 292: https://projecteuler.net/problem=292
package main

import (
	"fmt"
	"math"
	"sort"
)

const limit292 = 120

type vector292 struct {
	x int
	y int
	l int
}

func gcd292(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func isqrt292(n int) int {
	r := int(math.Sqrt(float64(n)))
	for (r+1)*(r+1) <= n {
		r++
	}
	for r*r > n {
		r--
	}
	return r
}

func pack292(x, y, p, c int) int {
	side := 2*limit292 + 1
	return (((p*side)+(x+limit292))*side+(y+limit292))*4 + c
}

func unpack292(k int) (int, int, int, int) {
	side := 2*limit292 + 1
	c := k % 4
	k /= 4
	y := k%side - limit292
	k /= side
	x := k%side - limit292
	p := k / side
	return x, y, p, c
}

func groups292(n int) [][]vector292 {
	byDir := make(map[[2]int][]vector292)

	for x := -n; x <= n; x++ {
		for y := -n; y <= n; y++ {
			if x == 0 && y == 0 {
				continue
			}

			l := isqrt292(x*x + y*y)
			if l*l != x*x+y*y || l > n {
				continue
			}

			g := gcd292(abs292(x), abs292(y))
			key := [2]int{x / g, y / g}
			byDir[key] = append(byDir[key], vector292{x, y, l})
		}
	}

	keys := make([][2]int, 0, len(byDir))
	for key := range byDir {
		keys = append(keys, key)
	}
	sort.Slice(keys, func(i, j int) bool {
		if keys[i][0] != keys[j][0] {
			return keys[i][0] < keys[j][0]
		}
		return keys[i][1] < keys[j][1]
	})

	groups := make([][]vector292, 0, len(keys))
	for _, key := range keys {
		v := byDir[key]
		sort.Slice(v, func(i, j int) bool {
			return v[i].l < v[j].l
		})
		groups = append(groups, v)
	}
	return groups
}

func abs292(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func count292(n int) int64 {
	dp := map[int]int64{pack292(0, 0, 0, 0): 1}

	for _, group := range groups292(n) {
		next := make(map[int]int64, len(dp))
		for k, v := range dp {
			next[k] = v
		}

		for k, ways := range dp {
			sx, sy, p, c := unpack292(k)
			for _, v := range group {
				np := p + v.l
				if np > n {
					continue
				}

				nc := c + 1
				if nc > 3 {
					nc = 3
				}
				next[pack292(sx+v.x, sy+v.y, np, nc)] += ways
			}
		}

		dp = next
	}

	var total int64
	for k, ways := range dp {
		x, y, _, c := unpack292(k)
		if x == 0 && y == 0 && c >= 3 {
			total += ways
		}
	}
	return total
}

func solve() int64 {
	return count292(limit292)
}

func main() {
	fmt.Println(solve())
}
